# Absenteeism at work project

by Andrea Mussap\
For more information, see [Github](link)

In [2]:
# Import libraries
import pandas as pd

In [3]:
# Import the dataset
df_raw_data = pd.read_csv('../dataset_raw/Absenteeism_data.csv')

## Section 1. Exploratory Data Analysis (EDA)

### Use `.head()` and `.tail()` functions to eyeball the columns and a few rows of the df

In [4]:
df_raw_data.head()

ID  Reason for Absence        Date  Transportation Expense  \
0  11                  26  07/07/2015                     289   
1  36                   0  14/07/2015                     118   
2   3                  23  15/07/2015                     179   
3   7                   7  16/07/2015                     279   
4  11                  23  23/07/2015                     289   

   Distance to Work  Age  Daily Work Load Average  Body Mass Index  Education  \
0                36   33                  239.554               30          1   
1                13   50                  239.554               31          1   
2                51   38                  239.554               31          1   
3                 5   39                  239.554               24          1   
4                36   33                  239.554               30          1   

   Children  Pets  Absenteeism Time in Hours  
0         2     1                          4  
1         1     0                          0  
2         0     0                          2  
3         2     0                          4  
4         2     1                          2

In [4]:
df_raw_data.tail()

ID  Reason for Absence        Date  Transportation Expense  \
695  17                  10  23/05/2018                     179   
696  28                   6  23/05/2018                     225   
697  18                  10  24/05/2018                     330   
698  25                  23  24/05/2018                     235   
699  15                  28  31/05/2018                     291   

     Distance to Work  Age  Daily Work Load Average  Body Mass Index  \
695                22   40                  237.656               22   
696                26   28                  237.656               24   
697                16   28                  237.656               25   
698                16   32                  237.656               25   
699                31   40                  237.656               25   

     Education  Children  Pets  Absenteeism Time in Hours  
695          2         2     0                          8  
696          1         1     2                          3  
697          2         0     0                          8  
698          3         0     0                          2  
699          1         1     1                          2

### Use the `shape` [attribute of pandas](https://pandas.pydata.org/pandas-docs/version/0.23.4/api.html#attributes-and-underlying-data) to get the dimensionality of the df

In [5]:
df_raw_data.shape

(700, 12)

The data frame comprises of 700 observations (rows) and 12 characteristics (columns)

### Check for a concise summary of the df (data types, null values)

In [6]:
df_raw_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 700 entries, 0 to 699
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   ID                         700 non-null    int64  
 1   Reason for Absence         700 non-null    int64  
 2   Date                       700 non-null    object 
 3   Transportation Expense     700 non-null    int64  
 4   Distance to Work           700 non-null    int64  
 5   Age                        700 non-null    int64  
 6   Daily Work Load Average    700 non-null    float64
 7   Body Mass Index            700 non-null    int64  
 8   Education                  700 non-null    int64  
 9   Children                   700 non-null    int64  
 10  Pets                       700 non-null    int64  
 11  Absenteeism Time in Hours  700 non-null    int64  
dtypes: float64(1), int64(10), object(1)
memory usage: 65.8+ KB


* The data frame contains integer, object, and float values.
* No null values

Use `describe` [function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html) to get basic statistics of a column.

In [7]:
df_raw_data['Absenteeism Time in Hours'].describe()

count    700.000000
mean       6.761429
std       12.670082
min        0.000000
25%        2.000000
50%        3.000000
75%        8.000000
max      120.000000
Name: Absenteeism Time in Hours, dtype: float64

This shows that employees were absent from work from 0 (min) to 120 (max) hours.

### `count` the `unique` values of a column, `sorting` by quantity

In [8]:
df_raw_data['Education'].value_counts(sort=True)

1    583
3     73
2     40
4      4
Name: Education, dtype: int64

From the dataset description:
* High school (1)
* Graduate (2)
* Postgraduate (3)
* Master/Doctor (4)

Meaning that around 83% of the data falls into the *High school* variable.

In [9]:
# check the reasons for absence
df_raw_data['Reason for Absence'].value_counts().sort_index()

0      38
1      16
2       1
3       1
4       2
5       3
6       6
7      13
8       5
9       4
10     22
11     24
12      8
13     52
14     18
15      2
16      3
17      1
18     21
19     36
21      6
22     32
23    147
24      3
25     29
26     31
27     66
28    110
Name: Reason for Absence, dtype: int64

In [10]:
df_raw_data['Transportation Expense'].min()

118

In [11]:
df_raw_data['Transportation Expense'].max()

388

There's a big difference between the `min` and `max` values of *Transportation Expense*.\
Since this study is about absenteeism from work due to medical reasons, transportation expenses\
shouldn't have an impact on the study. Should it?

### EDA Summary

This is a simple, high-level EDA. I've checked for data types, null values, differences among the values of the variables, and so on.\
This is a good start for the purpose of this exercise. Now, let's jump to the next part of the exercise - **Data transformation**.

## Section 2. Data Transformation

### Drop variables

Delete variables, which don't add value to the study,that can't help to explain the value of the dependent variable.\
I decided to create a new data frame, `df_clean_data`, for the transformations.

In [5]:
df_clean_data = df_raw_data.drop(['ID'], axis = 1)

In [13]:
display(df_clean_data)

Reason for Absence        Date  Transportation Expense  Distance to Work  \
0                    26  07/07/2015                     289                36   
1                     0  14/07/2015                     118                13   
2                    23  15/07/2015                     179                51   
3                     7  16/07/2015                     279                 5   
4                    23  23/07/2015                     289                36   
..                  ...         ...                     ...               ...   
695                  10  23/05/2018                     179                22   
696                   6  23/05/2018                     225                26   
697                  10  24/05/2018                     330                16   
698                  23  24/05/2018                     235                16   
699                  28  31/05/2018                     291                31   

     Age  Daily Work Load Average  Body Mass Index  Education  Children  Pets  \
0     33                  239.554               30          1         2     1   
1     50                  239.554               31          1         1     0   
2     38                  239.554               31          1         0     0   
3     39                  239.554               24          1         2     0   
4     33                  239.554               30          1         2     1   
..   ...                      ...              ...        ...       ...   ...   
695   40                  237.656               22          2         2     0   
696   28                  237.656               24          1         1     2   
697   28                  237.656               25          2         0     0   
698   32                  237.656               25          3         0     0   
699   40                  237.656               25          1         1     1   

     Absenteeism Time in Hours  
0                            4  
1                            0  
2                            2  
3                            4  
4                            2  
..                         ...  
695                          8  
696                          3  
697                          8  
698                          2  
699                          2  

[700 rows x 11 columns]

I could use [pandas options](https://pandas.pydata.org/pandas-docs/stable/user_guide/options.html) attributes to see the entire df, but I'm happy with the previous view.

### Split `Reason for absence` into multiple dummy variables

* For the reasoning about this step, see this [README.md](link.com) file

In [6]:
reason_columns = pd.get_dummies(df_clean_data['Reason for Absence'], drop_first = True)

In [35]:
reason_columns

1   2   3   4   5   6   7   8   9   10  ...  18  19  21  22  23  24  25  \
0     0   0   0   0   0   0   0   0   0   0  ...   0   0   0   0   0   0   0   
1     0   0   0   0   0   0   0   0   0   0  ...   0   0   0   0   0   0   0   
2     0   0   0   0   0   0   0   0   0   0  ...   0   0   0   0   1   0   0   
3     0   0   0   0   0   0   1   0   0   0  ...   0   0   0   0   0   0   0   
4     0   0   0   0   0   0   0   0   0   0  ...   0   0   0   0   1   0   0   
..   ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ...  ..  ..  ..  ..  ..  ..  ..   
695   0   0   0   0   0   0   0   0   0   1  ...   0   0   0   0   0   0   0   
696   0   0   0   0   0   1   0   0   0   0  ...   0   0   0   0   0   0   0   
697   0   0   0   0   0   0   0   0   0   1  ...   0   0   0   0   0   0   0   
698   0   0   0   0   0   0   0   0   0   0  ...   0   0   0   0   1   0   0   
699   0   0   0   0   0   0   0   0   0   0  ...   0   0   0   0   0   0   0   

     26  27  28  
0     1   0   0  
1     0   0   0  
2     0   0   0  
3     0   0   0  
4     0   0   0  
..   ..  ..  ..  
695   0   0   0  
696   0   0   0  
697   0   0   0  
698   0   0   0  
699   0   0   1  

[700 rows x 27 columns]

### Categorizing the various reasons for absence

Separate reason_columns data frame in 4 groups:

Group 1 - reasons 1-14. They're all related to various diseases.\
Group 2 - reasons 15-17. Reasons related to to pregnancy and giving birth.\
Group 3 - reasons 18-21. They're about poisoning or other reasons not elsewhere categorized.\
Group 4 - reasons 22 to the last. Categorized as light reasons.

We want to substitute an entire row of 14 values with a new single line showing: 

* 0 If *none* of the values on the given row were equal to 1
* 1 in case somewhere among these 14 columns we have *observed* the number one.

In [7]:
# test the command
reason_columns.loc[:, 1:14].max(axis=1)

0      0
1      0
2      0
3      1
4      0
      ..
695    1
696    1
697    1
698    0
699    0
Length: 700, dtype: uint8

In [8]:
# create the columns
reason_columns_1 = reason_columns.loc[:, 1:14].max(axis=1)
reason_columns_2 = reason_columns.loc[:, 15:17].max(axis=1)
reason_columns_3 = reason_columns.loc[:, 18:21].max(axis=1)
reason_columns_4 = reason_columns.loc[:, 22: ].max(axis=1)

In [13]:
reason_columns_1

0      0
1      0
2      0
3      1
4      0
      ..
695    1
696    1
697    1
698    0
699    0
Length: 700, dtype: uint8

### Checkpoint 1

Before adding reason_types to the data frame, delete 'Reason for Absence', but first create a checkpoint.\
In fact, I prefer to create a backup, which I can restore later, in case i need it. 

In [12]:
#df_backup1 = df_clean_data.copy()
#df_backup1

Date  Transportation Expense  Distance to Work  Age  \
0    07/07/2015                     289                36   33   
1    14/07/2015                     118                13   50   
2    15/07/2015                     179                51   38   
3    16/07/2015                     279                 5   39   
4    23/07/2015                     289                36   33   
..          ...                     ...               ...  ...   
695  23/05/2018                     179                22   40   
696  23/05/2018                     225                26   28   
697  24/05/2018                     330                16   28   
698  24/05/2018                     235                16   32   
699  31/05/2018                     291                31   40   

     Daily Work Load Average  Body Mass Index  Education  Children  Pets  \
0                    239.554               30          1         2     1   
1                    239.554               31          1         1     0   
2                    239.554               31          1         0     0   
3                    239.554               24          1         2     0   
4                    239.554               30          1         2     1   
..                       ...              ...        ...       ...   ...   
695                  237.656               22          2         2     0   
696                  237.656               24          1         1     2   
697                  237.656               25          2         0     0   
698                  237.656               25          3         0     0   
699                  237.656               25          1         1     1   

     Absenteeism Time in Hours  
0                            4  
1                            0  
2                            2  
3                            4  
4                            2  
..                         ...  
695                          8  
696                          3  
697                          8  
698                          2  
699                          2  

[700 rows x 10 columns]

In [70]:
#df_clean_data = df_backup1.copy()
#df_clean_data

Reason for Absence        Date  Transportation Expense  Distance to Work  \
0                    26  07/07/2015                     289                36   
1                     0  14/07/2015                     118                13   
2                    23  15/07/2015                     179                51   
3                     7  16/07/2015                     279                 5   
4                    23  23/07/2015                     289                36   
..                  ...         ...                     ...               ...   
695                  10  23/05/2018                     179                22   
696                   6  23/05/2018                     225                26   
697                  10  24/05/2018                     330                16   
698                  23  24/05/2018                     235                16   
699                  28  31/05/2018                     291                31   

     Age  Daily Work Load Average  Body Mass Index  Education  Children  Pets  \
0     33                  239.554               30          1         2     1   
1     50                  239.554               31          1         1     0   
2     38                  239.554               31          1         0     0   
3     39                  239.554               24          1         2     0   
4     33                  239.554               30          1         2     1   
..   ...                      ...              ...        ...       ...   ...   
695   40                  237.656               22          2         2     0   
696   28                  237.656               24          1         1     2   
697   28                  237.656               25          2         0     0   
698   32                  237.656               25          3         0     0   
699   40                  237.656               25          1         1     1   

     Absenteeism Time in Hours  
0                            4  
1                            0  
2                            2  
3                            4  
4                            2  
..                         ...  
695                          8  
696                          3  
697                          8  
698                          2  
699                          2  

[700 rows x 11 columns]

### End checkpoint 1

In [10]:
# Drop Reason for Absence:
df_clean_data = df_clean_data.drop(['Reason for Absence'], axis = 1)
df_clean_data

Date  Transportation Expense  Distance to Work  Age  \
0    07/07/2015                     289                36   33   
1    14/07/2015                     118                13   50   
2    15/07/2015                     179                51   38   
3    16/07/2015                     279                 5   39   
4    23/07/2015                     289                36   33   
..          ...                     ...               ...  ...   
695  23/05/2018                     179                22   40   
696  23/05/2018                     225                26   28   
697  24/05/2018                     330                16   28   
698  24/05/2018                     235                16   32   
699  31/05/2018                     291                31   40   

     Daily Work Load Average  Body Mass Index  Education  Children  Pets  \
0                    239.554               30          1         2     1   
1                    239.554               31          1         1     0   
2                    239.554               31          1         0     0   
3                    239.554               24          1         2     0   
4                    239.554               30          1         2     1   
..                       ...              ...        ...       ...   ...   
695                  237.656               22          2         2     0   
696                  237.656               24          1         1     2   
697                  237.656               25          2         0     0   
698                  237.656               25          3         0     0   
699                  237.656               25          1         1     1   

     Absenteeism Time in Hours  
0                            4  
1                            0  
2                            2  
3                            4  
4                            2  
..                         ...  
695                          8  
696                          3  
697                          8  
698                          2  
699                          2  

[700 rows x 10 columns]

In [14]:
# Add reason_types to the data frame
df_clean_data = pd.concat([df_clean_data, reason_columns_1, reason_columns_2,
                           reason_columns_3, reason_columns_4], axis = 1)
df_clean_data

Date  Transportation Expense  Distance to Work  Age  \
0    07/07/2015                     289                36   33   
1    14/07/2015                     118                13   50   
2    15/07/2015                     179                51   38   
3    16/07/2015                     279                 5   39   
4    23/07/2015                     289                36   33   
..          ...                     ...               ...  ...   
695  23/05/2018                     179                22   40   
696  23/05/2018                     225                26   28   
697  24/05/2018                     330                16   28   
698  24/05/2018                     235                16   32   
699  31/05/2018                     291                31   40   

     Daily Work Load Average  Body Mass Index  Education  Children  Pets  \
0                    239.554               30          1         2     1   
1                    239.554               31          1         1     0   
2                    239.554               31          1         0     0   
3                    239.554               24          1         2     0   
4                    239.554               30          1         2     1   
..                       ...              ...        ...       ...   ...   
695                  237.656               22          2         2     0   
696                  237.656               24          1         1     2   
697                  237.656               25          2         0     0   
698                  237.656               25          3         0     0   
699                  237.656               25          1         1     1   

     Absenteeism Time in Hours  0  1  2  3  
0                            4  0  0  0  1  
1                            0  0  0  0  0  
2                            2  0  0  0  1  
3                            4  1  0  0  0  
4                            2  0  0  0  1  
..                         ... .. .. .. ..  
695                          8  1  0  0  0  
696                          3  1  0  0  0  
697                          8  1  0  0  0  
698                          2  0  0  0  1  
699                          2  0  0  0  1  

[700 rows x 14 columns]

In [15]:
#df_backup2 = df_clean_data.copy()
#df_backup2

### Rename the new columns (0,1,2,3) with more meaningful names

We could use a dictionary to rename the columns, for example:\
`df_clean_data = df_clean_data.rename(columns={"Absenteeism Time in Hours": "Time"})`\
but, for the purpose of learning, let's use another technique: using a list.

In [24]:
# Retrieve a list with the column names
df_clean_data.columns.values

array(['Date', 'Transportation Expense', 'Distance to Work', 'Age',
       'Daily Work Load Average', 'Body Mass Index', 'Education',
       'Children', 'Pets', 'Absenteeism Time in Hours', 0, 1, 2, 3],
      dtype=object)

In [16]:
# Create a new list, and rename the 0,1,2,3 columns
column_names = (['Date', 'Transportation Expense', 'Distance to Work', 'Age',
       'Daily Work Load Average', 'Body Mass Index', 'Education',
       'Children', 'Pets', 'Absenteeism Time in Hours',
                 'Reason_1','Reason_2', 'Reason_3', 'Reason_4'])

In [76]:
column_names

['Date',
 'Transportation Expense',
 'Distance to Work',
 'Age',
 'Daily Work Load Average',
 'Body Mass Index',
 'Education',
 'Children',
 'Pets',
 'Absenteeism Time in Hours',
 'Reason_1',
 'Reason_2',
 'Reason_3',
 'Reason_4']

In [17]:
# Assign the list, the new names, to the columns of our data frame:
df_clean_data.columns = column_names
df_clean_data.head(3)

Date  Transportation Expense  Distance to Work  Age  \
0  07/07/2015                     289                36   33   
1  14/07/2015                     118                13   50   
2  15/07/2015                     179                51   38   

   Daily Work Load Average  Body Mass Index  Education  Children  Pets  \
0                  239.554               30          1         2     1   
1                  239.554               31          1         1     0   
2                  239.554               31          1         0     0   

   Absenteeism Time in Hours  Reason_1  Reason_2  Reason_3  Reason_4  
0                          4         0         0         0         1  
1                          0         0         0         0         0  
2                          2         0         0         0         1

In [78]:
#df_backup3 = df_clean_data.copy()
#df_backup3

Date  Transportation Expense  Distance to Work  Age  \
0    07/07/2015                     289                36   33   
1    14/07/2015                     118                13   50   
2    15/07/2015                     179                51   38   
3    16/07/2015                     279                 5   39   
4    23/07/2015                     289                36   33   
..          ...                     ...               ...  ...   
695  23/05/2018                     179                22   40   
696  23/05/2018                     225                26   28   
697  24/05/2018                     330                16   28   
698  24/05/2018                     235                16   32   
699  31/05/2018                     291                31   40   

     Daily Work Load Average  Body Mass Index  Education  Children  Pets  \
0                    239.554               30          1         2     1   
1                    239.554               31          1         1     0   
2                    239.554               31          1         0     0   
3                    239.554               24          1         2     0   
4                    239.554               30          1         2     1   
..                       ...              ...        ...       ...   ...   
695                  237.656               22          2         2     0   
696                  237.656               24          1         1     2   
697                  237.656               25          2         0     0   
698                  237.656               25          3         0     0   
699                  237.656               25          1         1     1   

     Absenteeism Time in Hours  Reason_1  Reason_2  Reason_3  Reason_4  
0                            4         0         0         0         1  
1                            0         0         0         0         0  
2                            2         0         0         0         1  
3                            4         1         0         0         0  
4                            2         0         0         0         1  
..                         ...       ...       ...       ...       ...  
695                          8         1         0         0         0  
696                          3         1         0         0         0  
697                          8         1         0         0         0  
698                          2         0         0         0         1  
699                          2         0         0         0         1  

[700 rows x 14 columns]

### Reorder columns

In [18]:
column_names_reordered = ['Reason_1','Reason_2', 'Reason_3', 'Reason_4',
                         'Date', 'Transportation Expense', 'Distance to Work', 'Age',
                         'Daily Work Load Average', 'Body Mass Index', 'Education',
                          'Children', 'Pets', 'Absenteeism Time in Hours']

In [19]:
# Assign the reordered list to our data frame:
df_clean_data = df_clean_data[column_names_reordered]
df_clean_data.head(3)

Reason_1  Reason_2  Reason_3  Reason_4        Date  Transportation Expense  \
0         0         0         0         1  07/07/2015                     289   
1         0         0         0         0  14/07/2015                     118   
2         0         0         0         1  15/07/2015                     179   

   Distance to Work  Age  Daily Work Load Average  Body Mass Index  Education  \
0                36   33                  239.554               30          1   
1                13   50                  239.554               31          1   
2                51   38                  239.554               31          1   

   Children  Pets  Absenteeism Time in Hours  
0         2     1                          4  
1         1     0                          0  
2         0     0                          2

## Transform column `Date`

Objective - Extract `Month` and `Day of week`\
This will allows us to check whether in specific months of the year, or days of the week,\
employees tend to be absent more often compared to other months or days.

In [85]:
df_clean_data['Date']

0      07/07/2015
1      14/07/2015
2      15/07/2015
3      16/07/2015
4      23/07/2015
          ...    
695    23/05/2018
696    23/05/2018
697    24/05/2018
698    24/05/2018
699    31/05/2018
Name: Date, Length: 700, dtype: object

The values on column `Date` follow the format:

* Day of the month
* Month
* Year

In [86]:
# Check the type of column 'Date'
type(df_clean_data['Date'])

pandas.core.series.Series

In [87]:
# Check the type of each value in this series
type(df_clean_data['Date'][0])

str

We know that in one column, or in one series, we can have values of a single data type only.\
The type() function shows that the values of this column are of type `string`. They are text.

In [90]:
#df_backup4 = df_clean_data.copy()
#df_backup4

Reason_1  Reason_2  Reason_3  Reason_4        Date  \
0           0         0         0         1  07/07/2015   
1           0         0         0         0  14/07/2015   
2           0         0         0         1  15/07/2015   
3           1         0         0         0  16/07/2015   
4           0         0         0         1  23/07/2015   
..        ...       ...       ...       ...         ...   
695         1         0         0         0  23/05/2018   
696         1         0         0         0  23/05/2018   
697         1         0         0         0  24/05/2018   
698         0         0         0         1  24/05/2018   
699         0         0         0         1  31/05/2018   

     Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                       289                36   33                  239.554   
1                       118                13   50                  239.554   
2                       179                51   38                  239.554   
3                       279                 5   39                  239.554   
4                       289                36   33                  239.554   
..                      ...               ...  ...                      ...   
695                     179                22   40                  237.656   
696                     225                26   28                  237.656   
697                     330                16   28                  237.656   
698                     235                16   32                  237.656   
699                     291                31   40                  237.656   

     Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  
0                 30          1         2     1                          4  
1                 31          1         1     0                          0  
2                 31          1         0     0                          2  
3                 24          1         2     0                          4  
4                 30          1         2     1                          2  
..               ...        ...       ...   ...                        ...  
695               22          2         2     0                          8  
696               24          1         1     2                          3  
697               25          2         0     0                          8  
698               25          3         0     0                          2  
699               25          1         1     1                          2  

[700 rows x 14 columns]

In [20]:
# convert 'Date' to timestamp type
df_clean_data['Date'] = pd.to_datetime(df_clean_data['Date'], format = '%d/%m/%Y')

In [30]:
df_clean_data['Date'] 

0     2015-07-07
1     2015-07-14
2     2015-07-15
3     2015-07-16
4     2015-07-23
         ...    
695   2018-05-23
696   2018-05-23
697   2018-05-24
698   2018-05-24
699   2018-05-31
Name: Date, Length: 700, dtype: datetime64[ns]

In [93]:
# verify that the type has changed to timestamp
type(df_clean_data['Date'][0])

pandas._libs.tslibs.timestamps.Timestamp

### Extract month

Objective - To create a new column, month, to know in which months people are more absent from work.\
We're going to create a list to collect all months, then append the list to the new column.

In [21]:
list_months = []

In [22]:
# The `for` loop reads the month of each row, and appends it to the list 
for i in range(df_clean_data.shape[0]):
       list_months.append(df_clean_data['Date'][i].month)

In [23]:
list_months

[7,
 7,
 7,
 7,
 7,
 7,
 7,
 7,
 7,
 7,
 7,
 7,
 7,
 7,
 7,
 7,
 7,
 7,
 7,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 10,
 10,
 10,
 10,
 10,
 10,
 10,
 10,
 10,
 10,
 10,
 10,
 10,
 10,
 10,
 10,
 10,
 10,
 11,
 11,
 11,
 11,
 11,
 11,
 11,
 11,
 11,
 11,
 11,
 11,
 11,
 11,
 11,
 11,
 11,
 11,
 12,
 12,
 12,
 12,
 12,
 12,
 12,
 12,
 12,
 12,
 12,
 12,
 12,
 12,
 12,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 6,
 6,
 6,
 6,
 6,
 6,
 6,
 6,
 6,
 6,
 6,
 6,
 6,
 6,
 6,
 6,
 7,
 

In [98]:
# check the length of the list
len(list_months)

700

In [24]:
# create a new column with the content of the list 
df_clean_data['Month Value'] = list_months

In [100]:
df_clean_data.head(20)

Reason_1  Reason_2  Reason_3  Reason_4       Date  Transportation Expense  \
0          0         0         0         1 2015-07-07                     289   
1          0         0         0         0 2015-07-14                     118   
2          0         0         0         1 2015-07-15                     179   
3          1         0         0         0 2015-07-16                     279   
4          0         0         0         1 2015-07-23                     289   
5          0         0         0         1 2015-07-10                     179   
6          0         0         0         1 2015-07-17                     361   
7          0         0         0         1 2015-07-24                     260   
8          0         0         1         0 2015-07-06                     155   
9          0         0         0         1 2015-07-13                     235   
10         1         0         0         0 2015-07-20                     260   
11         1         0         0         0 2015-07-14                     260   
12         1         0         0         0 2015-07-15                     260   
13         1         0         0         0 2015-07-15                     179   
14         0         0         0         1 2015-07-15                     179   
15         1         0         0         0 2015-07-17                     246   
16         0         0         0         1 2015-07-17                     179   
17         0         0         1         0 2015-07-27                     179   
18         1         0         0         0 2015-07-30                     189   
19         0         0         0         1 2015-08-05                     248   

    Distance to Work  Age  Daily Work Load Average  Body Mass Index  \
0                 36   33                  239.554               30   
1                 13   50                  239.554               31   
2                 51   38                  239.554               31   
3                  5   39                  239.554               24   
4                 36   33                  239.554               30   
5                 51   38                  239.554               31   
6                 52   28                  239.554               27   
7                 50   36                  239.554               23   
8                 12   34                  239.554               25   
9                 11   37                  239.554               29   
10                50   36                  239.554               23   
11                50   36                  239.554               23   
12                50   36                  239.554               23   
13                51   38                  239.554               31   
14                51   38                  239.554               31   
15                25   41                  239.554               23   
16                51   38                  239.554               31   
17                51   38                  239.554               31   
18                29   33                  239.554               25   
19                25   47                  205.917               32   

    Education  Children  Pets  Absenteeism Time in Hours  Month Value  
0           1         2     1                          4            7  
1           1         1     0                          0            7  
2           1         0     0                          2            7  
3           1         2     0                          4            7  
4           1         2     1                          2            7  
5           1         0     0                          2            7  
6           1         1     4                          8            7  
7           1         4     0                          4            7  
8           1         2     0                         40            7  
9           3         1     1                          8            7  
10          1     

### Extract day of the week

Use python [date.weekday()](https://docs.python.org/3/library/time.html) function to extract the day of the week as an integer.\
The approach for this task is to create a function, which returns the weekday from a given date.

In [25]:
def date_to_weekday(date_value):
    return date_value.weekday()

In [26]:
# create/ populate the new column, by calling the function
df_clean_data['Day of the Week'] = df_clean_data['Date'].apply(date_to_weekday)

In [27]:
df_clean_data.head()

Reason_1  Reason_2  Reason_3  Reason_4       Date  Transportation Expense  \
0         0         0         0         1 2015-07-07                     289   
1         0         0         0         0 2015-07-14                     118   
2         0         0         0         1 2015-07-15                     179   
3         1         0         0         0 2015-07-16                     279   
4         0         0         0         1 2015-07-23                     289   

   Distance to Work  Age  Daily Work Load Average  Body Mass Index  Education  \
0                36   33                  239.554               30          1   
1                13   50                  239.554               31          1   
2                51   38                  239.554               31          1   
3                 5   39                  239.554               24          1   
4                36   33                  239.554               30          1   

   Children  Pets  Absenteeism Time in Hours  Month Value  Day of the Week  
0         2     1                          4            7                1  
1         1     0                          0            7                1  
2         0     0                          2            7                2  
3         2     0                          4            7                3  
4         2     1                          2            7                3

In [28]:
df_clean_data = df_clean_data.drop(['Date'], axis = 1)

In [ ]:
df_clean_data.columns.values

In [29]:
reorder_date = ['Reason_1', 'Reason_2', 'Reason_3', 'Reason_4',
       'Month Value', 'Day of the Week',
       'Transportation Expense', 'Distance to Work', 'Age',
       'Daily Work Load Average', 'Body Mass Index', 'Education',
       'Children', 'Pets', 'Absenteeism Time in Hours']

In [30]:
# Assign the reordered list to our data frame:
df_clean_data = df_clean_data[reorder_date]
df_clean_data.head(3)

Reason_1  Reason_2  Reason_3  Reason_4  Month Value  Day of the Week  \
0         0         0         0         1            7                1   
1         0         0         0         0            7                1   
2         0         0         0         1            7                2   

   Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                     289                36   33                  239.554   
1                     118                13   50                  239.554   
2                     179                51   38                  239.554   

   Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  
0               30          1         2     1                          4  
1               31          1         1     0                          0  
2               31          1         0     0                          2

## Transform column `Education`

As we've seen in the EDA, more than 80% of the values in the `Education` variable falls under the High school category (1). Keeping the other three categories (2,3,4) separated doesn't seem quite relevant for this study, so it would make sense to combine them in a single category.\
The result should be (0) for High school, and (1) for any of the other degrees.\
The approach for this task is to overriding override the content of column `Education`, with the new values. We're going to do that by using a dictionary with key-value pairs

In [106]:
#df_backup5 = df_clean_data.copy()
#df_backup5

In [31]:
df_clean_data['Education'] = df_clean_data['Education'].map({1:0, 2:1, 3:1, 4:1})

In [108]:
df_clean_data['Education'].value_counts()

0    583
1    117
Name: Education, dtype: int64

In [34]:
df_clean_data

Reason_1  Reason_2  Reason_3  Reason_4  Month Value  Day of the Week  \
0           0         0         0         1            7                1   
1           0         0         0         0            7                1   
2           0         0         0         1            7                2   
3           1         0         0         0            7                3   
4           0         0         0         1            7                3   
..        ...       ...       ...       ...          ...              ...   
695         1         0         0         0            5                2   
696         1         0         0         0            5                2   
697         1         0         0         0            5                3   
698         0         0         0         1            5                3   
699         0         0         0         1            5                3   

     Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                       289                36   33                  239.554   
1                       118                13   50                  239.554   
2                       179                51   38                  239.554   
3                       279                 5   39                  239.554   
4                       289                36   33                  239.554   
..                      ...               ...  ...                      ...   
695                     179                22   40                  237.656   
696                     225                26   28                  237.656   
697                     330                16   28                  237.656   
698                     235                16   32                  237.656   
699                     291                31   40                  237.656   

     Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  
0                 30          0         2     1                          4  
1                 31          0         1     0                          0  
2                 31          0         0     0                          2  
3                 24          0         2     0                          4  
4                 30          0         2     1                          2  
..               ...        ...       ...   ...                        ...  
695               22          1         2     0                          8  
696               24          0         1     2                          3  
697               25          1         0     0                          8  
698               25          1         0     0                          2  
699               25          0         1     1                          2  

[700 rows x 15 columns]

## End of transformations

### Create the last checkpoint

In [35]:
df_preprocessed = df_clean_data.copy()
df_preprocessed.head()

Reason_1  Reason_2  Reason_3  Reason_4  Month Value  Day of the Week  \
0         0         0         0         1            7                1   
1         0         0         0         0            7                1   
2         0         0         0         1            7                2   
3         1         0         0         0            7                3   
4         0         0         0         1            7                3   

   Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                     289                36   33                  239.554   
1                     118                13   50                  239.554   
2                     179                51   38                  239.554   
3                     279                 5   39                  239.554   
4                     289                36   33                  239.554   

   Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  
0               30          0         2     1                          4  
1               31          0         1     0                          0  
2               31          0         0     0                          2  
3               24          0         2     0                          4  
4               30          0         2     1                          2

### Delete backups

In [32]:
del df_backup1

In [ ]:
# del df_backup2
# del df_backup3
# del df_backup4
# del df_backup5

## Export the new dataset

In [44]:
df_clean_data.to_csv('../dataset_new/Absenteeism_new.csv', sep=';')